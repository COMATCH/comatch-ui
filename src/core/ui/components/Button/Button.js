import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import classNames from 'classnames';

import { StyledWrapperButton, StyledWrapperLink } from './StyledWrapper';
// import './Button.scss';

const propTypes = {
    /**
     * Additional class names
     */
    className: PropTypes.string,
    disabled: PropTypes.bool,
    /**
     * Alternative styling with white background and primary colored text
     */
    ghost: PropTypes.bool,
    /**
     * If true, uses `<a>` instead of `<button>`
     */
    href: PropTypes.string,
    /**
     * HTML id attribute
     */
    id: PropTypes.string,
    /**
     * SVG icon
     */
    icon: PropTypes.node,
    iconAfterText: PropTypes.bool,
    onClick: PropTypes.func,
    /**
     * To be used with `href`, could set for example `target="_blank"`
     */
    target: PropTypes.string,
    /**
     * Text content of `Button`
     */
    text: PropTypes.string,
    type: PropTypes.oneOf(['button', 'submit', 'reset']),
    /**
     * Can have two values: rectangle(by dafauly) and circle. 
     */
    shape: PropTypes.oneOf(['rectangle', 'circle']),
};

const defaultProps = {
    className: '',
    disabled: false,
    ghost: false,
    href: '',
    id: null,
    icon: null,
    iconAfterText: false,
    onClick: noop,
    target: null,
    text: '',
    type: 'button',
    shape: 'rectangle',
};

function renderLinkButton({ classes, content, disabled, href, id, onClick, target, styledProps } = {}) {
    const finalOnClick = disabled ? (evt) => evt.preventDefault() : onClick;
    const rel = target && 'noopener noreferrer';
    return (
        <StyledWrapperLink
            {...styledProps}
            className={classes}
            href={href}
            target={target}
            id={id}
            rel={rel}
            onClick={finalOnClick}>
            {content}
        </StyledWrapperLink>
    );
}

function renderHtmlButton({ classes, content, id, onClick, type, styledProps } = {}) {
    return (
        <StyledWrapperButton
            {...styledProps}
            className={classes}
            id={id}
            type={type}
            onClick={onClick}>
            {content}
        </StyledWrapperButton>
    );
}

/**
 * A `Button` component with different stylistic variations. By default the styling is `full`,
 * but `ghost` is an alternative version with reversed colors.
 * 
 * If `href` is set an `<a>` is used instead of `<button>`. If the `target` prop is set to `_blank`
 * in combination with `href`, then it will also add the prop `rel="noopener noreferrer"` for
 * additional security.
 */

export const Button = ({
    id,
    className,
    href,
    target,
    onClick,

    ghost,
    disabled,

    text,
    type,
    icon,
    iconAfterText,
    shape,
}) => {
    const classes = classNames('Button', className, {
        full: !ghost,
        disabled,
        ghost,
        'only-icon': !text,
        'icon-after-text': text && iconAfterText,
    });
    const content = iconAfterText ? (
        <span>
            {text}
            {icon}
        </span>
    ) : (
        <span>
            {icon}
            {text}
        </span>
    );

    const calculatedProps = {
        id,
        classes,
        href,
        target,

        type,
        content,

        disabled,

        onClick,

        styledProps: {
            disabled,
            ghost,
            full: !ghost,
            onlyIcon: !text,
            iconAfterText: text && iconAfterText,
            shape,
        },
    };

    return href
        ? renderLinkButton(calculatedProps)
        : renderHtmlButton(calculatedProps);
};

Button.propTypes = propTypes;

Button.defaultProps = defaultProps;
