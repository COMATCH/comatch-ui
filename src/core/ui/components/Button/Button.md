
## Button

From [`Button.js`](Button.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**className** | `String` | `''` | :x: | 
**disabled** | `Boolean` | `false` | :x: | 
**ghost** | `Boolean` | `false` | :x: | 
**href** | `String` | `''` | :x: | 
**icon** | `ReactNode` | `null` | :x: | 
**iconAfterText** | `Boolean` | `false` | :x: | 
**id** | `String` | `null` | :x: | 
**onClick** | `Function` | `noop` | :x: | 
**target** | `String` | `null` | :x: | 
**text** | `String` | `''` | :x: | 
**type** | `Enum('button','submit','reset')` | `'button'` | :x: | 



