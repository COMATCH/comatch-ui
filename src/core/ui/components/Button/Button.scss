@import '../../../styles/variables.scss';

.Button {
    border: 1px solid;
    border-radius: 3px;
    display: inline-block;
    font-size: $co-font-sm;
    outline: 0 !important;
    font-size: $co-font-sm;
    touch-action: manipulation;
    cursor: pointer;
    text-align: center;
    text-transform: uppercase;
    white-space: nowrap;
    padding: 8px 15px; // makes the height 40px
    transition: background-color 250ms ease-out, border-color 250ms ease-out, color 250ms ease-out;

    &.full {
        background-color: $co-color-primary;
        border-color: $co-color-primary;
        color: $co-color-white;

        &.disabled {
            background-color: $co-color-light-gray !important;
        }

        &:hover {
            background-color: $co-color-secondary;
            border-color: $co-color-secondary;
        }
    }

    &.ghost {
        background-color: $co-color-white;
        border-color: $co-color-primary;
        color: $co-color-primary;

        &:hover {
            background-color: $co-color-secondary;
            border-color: $co-color-secondary;
            color: $co-color-white;
        }

        &.disabled {
            background-color: $co-color-white !important;
        }
    }

    &.disabled {
        border-color: $co-color-mid-gray !important;
        cursor: not-allowed;
        pointer-events: none;

        &,
        span {
            color: $co-color-mid-gray !important;
        }
    }

    // In case a Button has an icon defined
    svg {
        margin-right: 5px;
    }

    &.only-icon {
        svg {
            margin: 0;
        }
        padding: 8px 12px;
    }

    &.icon-after-text {
        white-space: initial;

        svg {
            float: right;
            position: relative;
            top: 2px;
            margin-left: 5px;
            margin-right: 0;
        }
    }

    // If buttons follow each other in a row
    // set a standardized whitespace between them
    & ~ & {
        margin-left: 9px;
    }
}
