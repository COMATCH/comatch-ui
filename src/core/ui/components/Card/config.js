const BACKGROUND_COLOR = {
    BLACK: 'black',
    WHITE: 'white',
    LIGHT_BLUE: 'lightBlue',
    PRIMARY: 'primary',
    LIGHT_PRIMARY: 'lightPrimary',
    SECONDARY: 'secondary',
    LIGHT_GRAY: 'lightGray',
    MID_GRAY: 'midGray',
    GRAY: 'gray',
    DARK_GRAY: 'darkGray',
    GREEN: 'green',
    LIGHT_GREEN: 'lightGreen',
    DARK_GREEN: 'darkGreen',
    RED: 'red',
    LIGHT_RED: 'lightRed',
    DARK_RED: 'darkRed',
    LIGHT_ORANGE: 'lightOrange',
    ORANGE: 'orange',
    DARK_ORANGE: 'darkOrange',
    YELLOW: 'yellow',
    TRANSPARENT: 'transparent',
};

const FONT_COLOR = {
    BLACK: 'white',
    WHITE: 'black',
    LIGHT_BLUE: 'black',
    PRIMARY: 'white',
    LIGHT_PRIMARY: 'black',
    SECONDARY: 'white',
    LIGHT_GRAY: 'black',
    MID_GRAY: 'black',
    GRAY: 'white',
    DARK_GRAY: 'white',
    GREEN: 'white',
    LIGHT_GREEN: 'black',
    DARK_GREEN: 'white',
    RED: 'white',
    LIGHT_RED: 'black',
    DARK_RED: 'white',
    LIGHT_ORANGE: 'black',
    ORANGE: 'black',
    DARK_ORANGE: 'white',
    YELLOW: 'black',
    TRANSPARENT: 'black',
};

const ORIENTATION = {
    HORIZONTAL: 'horizontal',
    VERTICAL: 'vertical',
};

export { BACKGROUND_COLOR, FONT_COLOR, ORIENTATION };
