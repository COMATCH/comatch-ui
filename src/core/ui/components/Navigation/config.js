export const ORIENTATION = {
    HORIZONTAL: 'horizontal',
    VERTICAL: 'vertical',
    RESPONSIVE: 'responsive',
};
