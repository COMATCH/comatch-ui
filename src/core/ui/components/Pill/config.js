import { palette, typography } from '../../../styles/variables';

const SIZES = {
    XS: 'XS',
    SM: 'SM',
    MD: 'MD',
    LG: 'LG',
    XL: 'XL',
    XXL: 'XXL',
};

const SIZES_IN_PIXELS = {
    XS: {
        fontSize: typography.fontXs,
        lineHeight: typography.lineHeightXs,
    },
    SM: {
        fontSize: typography.fontSm,
        lineHeight: typography.lineHeightSm,
    },
    MD: {
        fontSize: typography.fontMd,
        lineHeight: typography.lineHeightMd,
    },
    LG: {
        fontSize: typography.fontLg,
        lineHeight: typography.lineHeightLg,
    },
    XL: {
        fontSize: typography.fontXl,
        lineHeight: typography.lineHeightXl,
    },
    XXL: {
        fontSize: typography.fontXxl,
        lineHeight: typography.lineHeightXxl,
    },
};

const TYPES = {
    DISABLED: 'DISABLED',
    ERROR: 'ERROR',
    INFO: 'INFO',
    PRIMARY: 'PRIMARY',
    SECONDARY: 'SECONDARY',
    SUCCESS: 'SUCCESS',
    WARN: 'WARN',
};

const TYPES_IN_COLOR = {
    DISABLED: {
        background: palette.lightGray,
        borderColor: palette.lightGray,
        color: palette.black,
    },
    ERROR: {
        background: palette.red,
        borderColor: palette.red,
        color: palette.white,
    },
    INFO: {
        background: palette.lightBlue,
        borderColor: palette.lightBlue,
        color: palette.black,
    },
    PRIMARY: {
        background: palette.primary,
        borderColor: palette.primary,
        color: palette.white,
    },
    SECONDARY: {
        background: palette.secondary,
        borderColor: palette.secondary,
        color: palette.white,
    },
    SUCCESS: {
        background: palette.green,
        borderColor: palette.green,
        color: palette.white,
    },
    WARN: {
        background: palette.orange,
        borderColor: palette.orange,
        color: palette.white,
    },
};

export { SIZES, SIZES_IN_PIXELS, TYPES, TYPES_IN_COLOR };
