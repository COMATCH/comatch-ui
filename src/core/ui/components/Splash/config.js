import { typography } from '../../../styles/variables';

const ORIENTATION = {
    HORIZONTAL: 'horizontal',
    VERTICAL: 'vertical',
};

const SIZE = {
    XS: 'xs',
    SM: 'sm',
    MD: 'md',
    LG: 'lg',
    XL: 'xl',
    XXL: 'xxl',
};

const SIZE_TO_PIXELS = {
    xs: {
        title: `font-size: calc(${typography.fontXs} + 6px); line-height: calc(${typography.lineHeightXs} + 6px);`,
        subtitle: `font-size: ${typography.fontXs}; line-height: ${typography.fontXs};`,
    },
    sm: {
        title: `font-size: calc(${typography.fontSm} + 6px); line-height: calc(${typography.lineHeightSm} + 6px);`,
        subtitle: `font-size: ${typography.fontSm}; line-height: ${typography.fontSm};`,
    },
    md: {
        title: `font-size: calc(${typography.fontMd} + 6px); line-height: calc(${typography.lineHeightMd} + 6px);`,
        subtitle: `font-size: ${typography.fontMd}; line-height: ${typography.fontMd};`,
    },
    lg: {
        title: `font-size: calc(${typography.fontLg} + 6px); line-height: calc(${typography.lineHeightLg} + 6px);`,
        subtitle: `font-size: ${typography.fontLg}; line-height: ${typography.fontLg};`,
    },
    xl: {
        title: `font-size: calc(${typography.fontXl} + 6px); line-height: calc(${typography.lineHeightXl} + 6px);`,
        subtitle: `font-size: ${typography.fontXl}; line-height: ${typography.fontXl};`,
    },
    xxl: {
        title: `font-size: calc(${typography.fontXxl} + 6px); line-height: calc(${typography.lineHeightXxl} + 6px);`,
        subtitle: `font-size: ${typography.fontXxl}; line-height: ${typography.fontXxl};`,
    },
};

export { ORIENTATION, SIZE, SIZE_TO_PIXELS };
