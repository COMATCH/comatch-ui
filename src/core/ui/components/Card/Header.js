import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { StyledHeader, StyledHeaderControls } from './StyledWrapper';
import { BACKGROUND_COLOR } from './config';
import { Splash } from '../Splash';

const propTypes = {
    backgroundColor: PropTypes.oneOf(Object.keys(BACKGROUND_COLOR)),
    children: PropTypes.node,
    className: PropTypes.string,
    controls: PropTypes.node,
    id: PropTypes.string,
    title: Splash.propTypes.title,
    titleOrientation: Splash.propTypes.orientation,
    subtitle: Splash.propTypes.subtitle,
    wrap: PropTypes.bool,
};

const defaultProps = {
    backgroundColor: 'TRANSPARENT',
    children: null,
    className: null,
    controls: null,
    id: null,
    title: Splash.defaultProps.title,
    titleOrientation: Splash.defaultProps.orientation,
    subtitle: Splash.defaultProps.subtitle,
    wrap: false,
};

const Header = ({ backgroundColor, children, className, controls, id, title, titleOrientation, subtitle, wrap }) =>
    (!!children || !!title || !!subtitle || !!controls) && (
        <StyledHeader
            backgroundColor={backgroundColor}
            className={classnames('Header', className)}
            {...(id && { id })}
            wrap={wrap}
        >
            <Splash orientation={titleOrientation} title={title} subtitle={subtitle} />
            {children}
            <StyledHeaderControls>{controls}</StyledHeaderControls>
        </StyledHeader>
    );

Header.propTypes = propTypes;
Header.defaultProps = defaultProps;

export { Header };
