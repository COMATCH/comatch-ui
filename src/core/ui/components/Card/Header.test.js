import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Header } from './Header';
import { Button } from '../Button';
import { BACKGROUND_COLOR } from './config';
import '../../../config/tests/setup';

describe('Card.Header', () => {
    it('should render the default Header correctly', () => {
        const header = renderer.create(<Header>header</Header>).toJSON();
        expect(header).toMatchSnapshot();
    });

    it('should render a customized Header correctly', () => {
        const header = renderer
            .create(
                <Header
                    backgroundColor={BACKGROUND_COLOR.SECONDARY}
                    title="Header"
                    subtitle="Header subtitle"
                    controls={<Button>BTN</Button>}
                >
                    header
                </Header>,
            )
            .toJSON();
        expect(header).toMatchSnapshot();
    });

    it('should have the correct default props', () => {
        const header = mount(<Header />);
        expect(header.prop('backgroundColor')).toEqual('transparent');
        expect(header.prop('className')).toEqual(null);
        expect(header.prop('controls')).toEqual(null);
        expect(header.prop('id')).toEqual(null);
        expect(header.prop('title')).toEqual(null);
        expect(header.prop('titleOrientation')).toEqual('vertical');
        expect(header.prop('subtitle')).toEqual(null);
        expect(header.prop('wrap')).toEqual(false);
    });

    it('should have the correct nodes (corresponding to props)', () => {
        const emptyHeader = mount(<Header />);
        const defaultHeader = mount(<Header>header</Header>);
        const header = mount(
            <Header
                backgroundColor={BACKGROUND_COLOR.SECONDARY}
                title="Header"
                subtitle="Header subtitle"
                controls={<Button>BTN</Button>}
            >
                header
            </Header>,
        );

        // 1. Validate the props:
        expect(header.prop('backgroundColor')).toEqual('secondary');
        expect(header.prop('title')).toEqual('Header');
        expect(header.prop('subtitle')).toEqual('Header subtitle');
        expect(header.prop('controls')).toEqual(<Button>BTN</Button>);

        // 2. Validate the DOM elements:
        expect(header.find('.Splash.TitleAndSubtitle')).toBeTruthy();
        expect(header.find('.Children')).toBeTruthy();
        expect(header.find('.Controls')).toBeTruthy();

        expect(defaultHeader.find('.Splash.TitleAndSubtitle')).toBeTruthy();
        expect(defaultHeader.find('.Children')).toBeTruthy();
        expect(defaultHeader.find('.Controls')).toHaveLength(0);

        expect(emptyHeader.find('.Splash.TitleAndSubtitle')).toHaveLength(0);
        expect(emptyHeader.find('.Children')).toHaveLength(0);
        expect(emptyHeader.find('.Controls')).toHaveLength(0);
    });
});
