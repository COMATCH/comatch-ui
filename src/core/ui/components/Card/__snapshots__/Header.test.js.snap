// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.Header should render a customized Header correctly 1`] = `
<div
  className="Header sc-bxivhb iczkCS"
>
  <div
    className="Splash TitleAndSubtitle sc-gzVnrw bpmeZP"
  >
    <div
      className="Title sc-bZQynM bPbNzY"
    >
      Header
    </div>
    <div
      className="Subtitle sc-EHOje jUaBYT"
    >
      Header subtitle
    </div>
  </div>
  <div
    className="Children sc-bwzfXH uTxCW"
  >
    header
  </div>
  <div
    className="Controls sc-htpNat cPrsnq"
  >
    <button
      className="Button rectangle full only-icon sc-iwsKbI dlgvgE"
      disabled={false}
      id={null}
      onClick={[Function]}
      type="button"
    >
      <span>
        
      </span>
    </button>
  </div>
</div>
`;

exports[`Card.Header should render the default Header correctly 1`] = `
<div
  className="Header sc-bxivhb iczkCS"
>
  <div
    className="Children sc-bwzfXH uTxCW"
  >
    header
  </div>
</div>
`;
